### ISSUE TITLE
Allow unit tests to modify the Project created by the Verify methods

### ISSUE THREAD

**Issue by lgolding** [dotnet/roslyn-analyzers#881]
**Motivation**: I'm writing analyzer CA1812, AvoidUninstantiatedInternalClasses. The idea is if you have an internal class, and nobody in the assembly it lives in ever instantiates it, then it's useless, because nobody outside of the assembly can see it to instantiate it.

There are certain exceptions; for example, we don't report a diagnostic if the class contains the program entry point:

```
internal class Program
{
    private static void Main() {}
}
```

If this code occurs in a .exe, it is the program entry point, and we don't report it. But if it occurs in a .dll, it can't be the program entry point, and we do report it.

The problem is, you can't unit test this. The unit test infrastructure supplies a set of `static` `Verify` methods on the `DiagnosticAnalyzerTestBase` class. Ultimately, these methods create a `Project` with a default set of compilation options (specifically, `OutputKind.DynamicallyLinkedLibrary`), and there's no way to override this (because the methods are `static`).

To fix this, we'd either have to make all those methods instance methods and supply some overrides, or we'd have to plumb another, optional parameter down through the entire infrastructure, starting at the `Verify` methods. But that's not possible, given the current signature of the `Verify` methods, which is

```
protected void VerifyCSharp(string source, params DiagnosticResult[] expected)
```

Since the last parameter is a `params`, we can't add anything after it, and adding a parameter before it would break every test in the world.

**Comment by mavasani** [comment]
You can just define these types in source for your tests - just like the tests in the above file that failed when you made the change

**Comment by lgolding** [comment]
Yes, @srivatsn suggested that too. I'm making that change now.
